### PR TITLE
[MM:55208] tilde ~ and at @ cause searches to return fewer results than expected

### PR DESCRIPTION
**Pull request by ludvigbolin**
#### Summary

Before, adding `~` to a channel name in the `in:` field would cause nothing to match. Adding `@` to a username in the `from:` field would cause nothing to match.

Now, these are both adjusted and thereby returns the expected results.

You can now name a channel `~my-team`, and it will be found with searching `in:~my-team`.
If you search for a user using `from:@new-user`, messages from `new-user` will appear.

#### Ticket Link

Fixes: https://github.com/mattermost/mattermost/issues/25354
Jira: https://mattermost.atlassian.net/browse/MM-55208

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Comment by lieut-data** [comment]
Thanks @ludvigbolin & @hanzei! I won't be able to review this one for a while, so I've looped myself out of the review for now. @agarciamontoro, maybe another XYZ team member would be in a position to take a look?

File: server/channels/app/post.go [hunk 1]
```diff
@@ -1447,7 +1447,12 @@ func (a *App) parseAndFetchChannelIdByNameFromInFilter(c request.CTX, channelNam
 		if err != nil {
 			return nil, err
 		}
-		return channel, nil
+		if (channel != nil) {
+			return channel, nil
+		}
+	}
+	if strings.HasPrefix(channelName, "~") || strings.HasPrefix(channelName, "@") {
+		channelName = channelName[1:]
```
Comment (agarciamontoro) [inline server/channels/app/post.go:1455]: I don't think it makes sense to allow channels prefixed by `@`, so we can leave the `return channel, nil` as it was and then, either simply check by `strings.HasPrefix(channelName, "~")`, or do `channelName = strings.TrimPrefix(channelName, "@")` as you do with users below.

File: server/channels/app/post.go [hunk 2]
```diff
@@ -1510,6 +1515,7 @@ func (a *App) convertChannelNamesToChannelIds(c request.CTX, channels []string,
 
 func (a *App) convertUserNameToUserIds(usernames []string) []string {
 	for idx, username := range usernames {
+		username = strings.TrimPrefix(username, `@`)
```
Comment (agarciamontoro) [inline server/channels/app/post.go:1518]: We can use double quotes here
```suggestion
		username = strings.TrimPrefix(username, "@")
```

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
Thanks, @ludvigbolin! I do think this improves things! And while the changes make sense to me, and code-wise look good (except for the minor comments I left below), I don't know if there's something we should polish UX-wise. Looping @sbishel and @asaadmahmood in to review as well.

File: server/channels/app/post.go (hunk 1 above)
Comment (ludvigbolin) [reply server/channels/app/post.go]: Yeah that's reasonable. I've changed it now!

**Review by agarciamontoro — APPROVED** [review]
Thanks! I think this is more than reasonable. We'll need some thorough QA here apart from Scott's team input, since it's a hot code path, but other than that, this looks great to me! Thanks, @ludvigbolin!

**Review by sbishel — CHANGES_REQUESTED** [review]
Code looks good, but lets add a unit test for this.

**Comment by ludvigbolin** [comment]
> Code looks good, but lets add a unit test for this.

Should I do this by adding onto `TestSearchPostsForUser()` or should I create a new test that only checks the `~` and `@`? I find it a bit difficult to understand the test `TestSearchPostsForUser()` regarding how ElasticSearch is used. Do I need to do the same thing in my case or can I only create some dummy posts and check that `SearchPostsForUser()` will find these posts? Thanks for the feedback!

**Comment by ludvigbolin** [comment]
Hi! 
I am currently trying to add a new unit test but am running into some issues. My idea is to change the `searchTerm` which is used in `TestSearchPostsForUser` to include `"in:~" + fmt.Sprint(th.BasicChannel.Id)` but the method `GetChannelByName()` always returns an error as if there exists no channel. The same thing happens if I add `"in:" + fmt.Sprint(th.BasicChannel.Id)` to the searchTerm, so in a way the tilde does not cause this but is something about how the channel is set up. Perhaps @sbishel might know more about how the test channels are set up?

**Comment by ludvigbolin** [comment]
After some struggle, I managed to create a unit test that I believe to test the purpose correctly. Would you please have a look @sbishel ? Thanks!

**Comment by ludvigbolin** [comment]
> Hi! I am currently trying to add a new unit test but am running into some issues. My idea is to change the `searchTerm` which is used in `TestSearchPostsForUser` to include `"in:~" + fmt.Sprint(th.BasicChannel.Id)` but the method `GetChannelByName()` always returns an error as if there exists no channel. The same thing happens if I add `"in:" + fmt.Sprint(th.BasicChannel.Id)` to the searchTerm, so in a way the tilde does not cause this but is something about how the channel is set up. Perhaps @sbishel might know more about how the test channels are set up?

I fixed this by using `"in:~" + fmt.Sprint(th.BasicChannel.Name) + " " + searchTerm`, which made a lot more sense

**Review by sbishel — CHANGES_REQUESTED** [review]
Yep..that unit test should work. Let's add one more test to test the username changes as well. Should essentially be the same, except testing the "from:" query. 

But looks good. We'll get QA to make sure nothing else breaks.

**Review by lindy65 — APPROVED** [review]
Thanks @ludvigbolin - looks good to me 👍

**Review by asaadmahmood — CHANGES_REQUESTED** [review]
Is it possible to keep the autocomplete open if a person presses the tilda?
https://www.loom.com/share/dc8c8f00ffee4a14a23ab5b33d4d1f09

See video above.

**Comment by hloeung** [comment]
Would this be better done in `model.ParseSearchParams()`?

So I'm thinking https://github.com/mattermost/mattermost/blob/master/server/public/model/search_params.go#L274C1-L286C5:

```diff
--- search_params.go.orig       2023-12-20 12:10:36.097175559 +1100
+++ search_params.go    2023-12-20 12:11:51.057814553 +1100
@@ -273,16 +273,18 @@ func ParseSearchParams(text string, time

        for _, flag := range flags {
                if flag.name == "in" || flag.name == "channel" {
+                       channelName := strings.TrimPrefix(flag.value, "~")
                        if flag.exclude {
-                               excludedChannels = append(excludedChannels, flag.value)
+                               excludedChannels = append(excludedChannels, channelName)
                        } else {
-                               inChannels = append(inChannels, flag.value)
+                               inChannels = append(inChannels, channelName)
                        }
                } else if flag.name == "from" {
+                       username := strings.TrimPrefix(flag.value, "@")
                        if flag.exclude {
-                               excludedUsers = append(excludedUsers, flag.value)
+                               excludedUsers = append(excludedUsers, username)
                        } else {
-                               fromUsers = append(fromUsers, flag.value)
+                               fromUsers = append(fromUsers, username)
                        }
                } else if flag.name == "after" {
                        if flag.exclude {
```

**Comment by Willyfrog** [comment]
closing as https://github.com/mattermost/mattermost/pull/28146 was merged

**Pull request closed without merging**